src/algebra/ring/basic.lean

```diff
@@ -634,6 +639,16 @@ class ring (α : Type u) extends add_comm_group α, monoid α, distrib α
 section ring
 variables [ring α] {a b c d e : α}
 
+/- A (unital, associative) ring is a not-necessarily-unital, not-necessarily-associative ring -/
+@[priority 100] -- see Note [lower instance priority]
+instance ring.to_non_unital_non_assoc_ring :
+  non_unital_non_assoc_ring α :=
+{ zero_mul := λ a, add_left_cancel $ show 0 * a + 0 * a = 0 * a + 0,
+    by rw [← add_mul, zero_add, add_zero],
+  mul_zero := λ a, add_left_cancel $ show a * 0 + a * 0 = a * 0 + 0,
+    by rw [← mul_add, add_zero, add_zero],
+  ..‹ring α› }
```

Comment:
You should be able to use `..ring.to_non_unital_non_assoc_ring` in place of lines 658-661 below, to avoid duplicating the same proof.

Reply:
I've tried that - but like my other attempts to avoid this duplication, it leads to deterministic timeouts: https://github.com/leanprover-community/mathlib/runs/4674124721?check_suite_focus=true

Reply:
I think this has now been solved by https://github.com/leanprover-community/mathlib/pull/11124/commits/8c62c331528eec62815ecec20c113b756cc61797